### DISCUSSION TITLE
Download file from Public Folder

### DISCUSSION THREAD

**Discussion by theinfosecguy** [vercel/next.js#30751] [Help]
Hey All,

I am using the API route to process a CSV file using a python script and save the output CSV file to the public folder. I want to download the file once it is saved.

While using it on the local machine, it worked fine (saved + downloaded) but when deployed to EC2, the files are saved successfully to the public folder but I am not able to download that.

Any help would be appreciated.

**Comment by wwwdepot** [comment]
> Only assets that are in the public directory at build time will be served by Next.js. Files added at runtime won't be available.
> https://nextjs.org/docs/basic-features/static-file-serving

1. Save your files to S3 or external hosts. 
2. If you don't need storage, process the file & sent it right away.
3. Or if you are using Nginx, you could add a location to the `public` folder

```nginx
# Next.js
location / {
    proxy_pass http://localhost:3000; 
}

# Create a location that doesn't conflict with Next.js
# GET `/some-path/processed.csv`,  `/var/www/nextjs-app/public/processed.csv` will be sent
# >> http://nginx.org/en/docs/http/ngx_http_core_module.html#alias
location /some-path/ {
  alias /var/www/nextjs-app/public/;
  expires -1;
  default_type application/octet-stream;
}
```

**Comment by theinfosecguy** [comment]
Can you elaborate the 2nd option?

Also, The folder and the file is dynamic. How to handle that with NGINX?

Thanks for the help :)

**Reply by wwwdepot** [reply]
2nd option is technically serving the CSV file on-the-fly

1. API route receives a new request
2. Python script executed, output saved to a tmp file in OS tmp directory.
3. API route reads the tmp file from above, whether or not cleanup tmp file is up to you.
4. API route returns response, browser downloads csv file.

(If you're using STDOUT, you can skip the save/read tmp file steps)

NGINX is a great tool for serving static files, put it simple, if NGINX can read static files on disk, it can serve it.
